File: python/samples/advanced_local_ids.py
#!/usr/bin/env python3
"""Advanced Local ID Operations - Vista SDK Python.

This example demonstrates advanced Local ID operations:
- Building complex Local IDs with multiple components
- Error handling during parsing
- Working with custom tags
- Local ID validation and inspection
"""

from vista_sdk.codebook_names import CodebookName
from vista_sdk.gmod_path import GmodPath
from vista_sdk.local_id_builder import LocalIdBuilder
from vista_sdk.local_id_builder_parsing import LocalIdBuilderParsing
from vista_sdk.vis import VIS
from vista_sdk.vis_version import VisVersion


def main() -> None:  # noqa : C901
    """Demonstrate advanced Local ID operations."""
    print("=== Advanced Local ID Operations ===\n")

    # Initialize VIS
    vis = VIS()
    version = VisVersion.v3_4a
    codebooks = vis.get_codebooks(version)

    # 1. Building complex Local IDs
    print("1. Building Complex Local IDs...")

    # Primary and secondary items
    primary_path = GmodPath.parse("411.1/C101.31-2", arg=version)
    secondary_path = GmodPath.parse("411.1/C101.63/S206", arg=version)

    complex_local_id = (
        LocalIdBuilder.create(version)
        .with_primary_item(primary_path)
        .with_secondary_item(secondary_path)
        .with_metadata_tag(codebooks.create_tag(CodebookName.Quantity, "temperature"))
        .with_metadata_tag(codebooks.create_tag(CodebookName.Content, "exhaust.gas"))
        .with_metadata_tag(codebooks.create_tag(CodebookName.State, "high"))
        .with_metadata_tag(codebooks.create_tag(CodebookName.Position, "inlet"))
        .build()
    )

    print(f"   Complex Local ID: {complex_local_id}")
    print(f"   Has secondary item: {complex_local_id.secondary_item is not None}")
    print(f"   Number of metadata tags: {len(complex_local_id.metadata_tags)}")

    # 2. Working with custom tags
    print("\n2. Working with Custom Tags...")

    # Create custom tags (prefix with ~)
    custom_quantity = codebooks.try_create_tag(
        CodebookName.Quantity, "custom_temperature"
    )
    custom_position = codebooks.try_create_tag(CodebookName.Position, "custom_location")

    if custom_quantity and custom_position:
        custom_local_id = (
            LocalIdBuilder.create(version)
            .with_primary_item(primary_path)
            .with_metadata_tag(custom_quantity)
            .with_metadata_tag(custom_position)
            .build()
        )

        print(f"   Custom Local ID: {custom_local_id}")
        print(f"   Quantity tag is custom: {custom_quantity.is_custom}")
        print(f"   Position tag is custom: {custom_position.is_custom}")

    # 3. Error handling and validation
    print("\n3. Error Handling and Validation...")

    invalid_local_ids = [
        "/dnv-v2/vis-3-4a/invalid-path/meta/qty-temperature",
        "/invalid-naming-rule/vis-3-4a/411.1/meta/qty-temperature",
        "/dnv-v2/vis-3-4a/411.1/meta/qty-invalid_quantity",
        # Invalid position order
        "/dnv-v2/vis-3-4a/411.1/meta/qty-temperature/pos-1-centre",
    ]

    for invalid_id in invalid_local_ids:
        print(f"\n   Testing invalid Local ID: {invalid_id}")

        try:
            success, errors, result = LocalIdBuilder.try_parse(invalid_id)

            if not success or result is None:
                print("     ✗ Parsing failed as expected")
                if errors.has_errors:
                    print("     Errors found:")
                    for error_type, message in errors:
                        print(f"       - {error_type}: {message}")
            else:
                print(f"     ⚠ Unexpectedly parsed successfully: {result}")

        except Exception as e:
            print(f"     ✗ Exception during parsing: {e}")

    # 4. Local ID inspection and analysis
    print("\n4. Local ID Inspection and Analysis...")

    valid_local_ids = [
        "/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature/cnt-exhaust.gas/pos-inlet",
        "/dnv-v2/vis-3-4a/652.31/S90.3/S61/sec/652.1i-1P/meta/cnt-sea.water/state-opened",
        "/dnv-v2/vis-3-4a/1021.1i-6P/H123/meta/qty-volume/cnt-cargo/pos~percentage",
    ]

    for local_id_str in valid_local_ids:
        print(f"\n   Analyzing: {local_id_str}")

        try:
            parser = LocalIdBuilderParsing()
            local_id = parser.parse(local_id_str).build()

            print("     ✓ Successfully parsed and built")
            print(f"     ✓ VIS Version: {local_id.vis_version}")
            print(f"     ✓ Has custom tags: {local_id.has_custom_tag}")
            print(f"     ✓ Primary item: {local_id.primary_item}")

            if local_id.secondary_item:
                print(f"     ✓ Secondary item: {local_id.secondary_item}")

            # Analyze each metadata tag
            print("     ✓ Metadata tags:")
            for tag in local_id.metadata_tags:
                print(
                    f"       - {tag.name.name}: {tag.value} (custom: {tag.is_custom})"
                )

        except Exception as e:
            print(f"     ✗ Failed to parse: {e}")

    # 5. Builder pattern variations
    print("\n5. Builder Pattern Variations...")

    # Using try_with methods for safe building
    builder = LocalIdBuilder.create(version)

    # Try adding components safely
    success, path_for_builder = GmodPath.try_parse("411.1/C101.31", version)
    if success and path_for_builder:
        builder = builder.with_primary_item(path_for_builder)

    # Try with valid tags
    temp_tag = codebooks.try_create_tag(CodebookName.Quantity, "temperature")
    if temp_tag:
        builder = builder.with_metadata_tag(temp_tag)

    exhaust_tag = codebooks.try_create_tag(CodebookName.Content, "exhaust.gas")
    if exhaust_tag:
        builder = builder.with_metadata_tag(exhaust_tag)

    # Try with potentially invalid values (this should fail)
    invalid_tag = codebooks.try_create_tag(
        CodebookName.Position, "invalid-position-format"
    )
    if invalid_tag:
        builder = builder.with_metadata_tag(invalid_tag)

    try:
        safe_local_id = builder.build()
        print(f"   Safe building result: {safe_local_id}")
    except Exception as e:
        print(f"   Safe building failed: {e}")

    # 6. Verbose mode demonstration
    print("\n6. Verbose Mode...")

    verbose_local_id = (
        LocalIdBuilder.create(version)
        .with_verbose_mode(True)
        .with_primary_item(primary_path)
        .with_metadata_tag(codebooks.create_tag(CodebookName.Quantity, "temperature"))
        .build()
    )

    regular_local_id = (
        LocalIdBuilder.create(version)
        .with_verbose_mode(False)
        .with_primary_item(primary_path)
        .with_metadata_tag(codebooks.create_tag(CodebookName.Quantity, "temperature"))
        .build()
    )

    print(f"   Verbose mode: {verbose_local_id}")
    print(f"   Regular mode: {regular_local_id}")

    print("\n=== Advanced operations completed! ===")


if __name__ == "__main__":
    main()

File: js/samples/AdvancedLocalIds.ts
/*
    Advanced Local ID Operations - Vista SDK JS/TS

    This example demonstrates advanced Local ID operations:
    - Building complex Local IDs with multiple components
    - Error handling during parsing
    - Working with custom tags
    - Local ID validation and inspection
*/

import {
    CodebookName,
    LocalId,
    LocalIdBuilder,
    LocalIdParsingErrorBuilder,
    ParsingState,
    VIS,
    VisVersion,
    VisVersionExtension,
} from "dnv-vista-sdk";

async function main() {
    console.log("=== Advanced Local ID Operations ===\n");

    const version = VisVersion.v3_4a;
    const vis = VIS.instance;
    const { gmod, codebooks, locations } = await vis.getVIS(version);

    // 1. Building complex Local IDs
    console.log("1. Building Complex Local IDs...");

    const primaryPath = gmod.parsePath("411.1/C101.31-2", locations);
    const secondaryPath = gmod.parsePath("411.1/C101.63/S206", locations);

    const complexLocalId = LocalIdBuilder.create(version)
        .withPrimaryItem(primaryPath)
        .withSecondaryItem(secondaryPath)
        .withMetadataTag(
            codebooks.createTag(CodebookName.Quantity, "temperature"),
        )
        .withMetadataTag(
            codebooks.createTag(CodebookName.Content, "exhaust.gas"),
        )
        .withMetadataTag(codebooks.createTag(CodebookName.State, "high"))
        .withMetadataTag(codebooks.createTag(CodebookName.Position, "inlet"))
        .build();

    console.log(`   Complex Local ID: ${complexLocalId}`);
    console.log(
        `   Has secondary item: ${complexLocalId.secondaryItem !== undefined}`,
    );
    console.log(
        `   Number of metadata tags: ${complexLocalId.metadataTags.length}`,
    );

    // 2. Working with custom tags
    console.log("\n2. Working with Custom Tags...");

    const customQuantity = codebooks.tryCreateTag(
        CodebookName.Quantity,
        "custom_temperature",
    );
    const customPosition = codebooks.tryCreateTag(
        CodebookName.Position,
        "custom_location",
    );

    if (customQuantity && customPosition) {
        const customLocalId = LocalIdBuilder.create(version)
            .withPrimaryItem(primaryPath)
            .withMetadataTag(customQuantity)
            .withMetadataTag(customPosition)
            .build();

        console.log(`   Custom Local ID: ${customLocalId}`);
        console.log(
            `   Quantity tag is custom: ${customLocalId.builder.quantity?.isCustom}`,
        );
        console.log(
            `   Position tag is custom: ${customLocalId.builder.position?.isCustom}`,
        );
    }

    // 3. Error handling and validation
    console.log("\n3. Error Handling and Validation...");

    const invalidLocalIds = [
        "/dnv-v2/vis-3-4a/invalid-path/meta/qty-temperature",
        "/invalid-naming-rule/vis-3-4a/411.1/meta/qty-temperature",
        "/dnv-v2/vis-3-4a/411.1/meta/qty-invalid_quantity",
    ];

    for (const invalidId of invalidLocalIds) {
        console.log(`\n   Testing invalid Local ID: ${invalidId}`);
        const errorBuilder = new LocalIdParsingErrorBuilder();
        const result = LocalIdBuilder.tryParse(
            invalidId,
            gmod,
            codebooks,
            locations,
            errorBuilder,
        );

        if (!result) {
            console.log("     ✗ Parsing failed as expected");
            if (errorBuilder.hasError) {
                console.log("     Errors found:");
                for (const e of errorBuilder.errors) {
                    console.log(
                        `       - ${ParsingState[e.type]}: ${e.message}`,
                    );
                }
            }
        } else {
            console.log(`     ⚠ Unexpectedly parsed successfully: ${result}`);
        }
    }

    // 4. Local ID inspection and analysis
    console.log("\n4. Local ID Inspection and Analysis...");

    const validLocalIds = [
        "/dnv-v2/vis-3-4a/411.1/C101.31-2/meta/qty-temperature/cnt-exhaust.gas/pos-inlet",
        "/dnv-v2/vis-3-4a/652.31/S90.3/S61/sec/652.1i-1P/meta/cnt-sea.water/state-opened",
        "/dnv-v2/vis-3-4a/1021.1i-6P/H123/meta/qty-volume/cnt-cargo/pos~percentage",
    ];

    for (const localIdStr of validLocalIds) {
        console.log(`\n   Analyzing: ${localIdStr}`);

        try {
            const localId = await LocalId.parseAsync(localIdStr);
            console.log("     ✓ Successfully parsed");
            console.log(
                `     ✓ VIS Version: ${VisVersionExtension.toVersionString(localId.visVersion)}`,
            );
            console.log(`     ✓ Has custom tags: ${localId.hasCustomTag}`);
            console.log(`     ✓ Primary item: ${localId.primaryItem}`);

            if (localId.secondaryItem) {
                console.log(`     ✓ Secondary item: ${localId.secondaryItem}`);
            }

            console.log("     ✓ Metadata tags:");
            for (const tag of localId.metadataTags) {
                console.log(
                    `       - ${CodebookName[tag.name]}: ${tag.value} (custom: ${tag.isCustom})`,
                );
            }
        } catch {
            console.log("     ✗ Failed to parse");
        }
    }

    // 5. Builder pattern variations
    console.log("\n5. Builder Pattern Variations...");

    let builder = LocalIdBuilder.create(version);

    // Try adding components safely
    const pathForBuilder = gmod.tryParsePath("411.1/C101.31", locations);
    if (pathForBuilder) {
        builder = builder.withPrimaryItem(pathForBuilder);
    }

    // Try with valid tags
    const tempTag = codebooks.tryCreateTag(
        CodebookName.Quantity,
        "temperature",
    );
    if (tempTag) {
        builder = builder.withMetadataTag(tempTag);
    }

    const exhaustTag = codebooks.tryCreateTag(
        CodebookName.Content,
        "exhaust.gas",
    );
    if (exhaustTag) {
        builder = builder.withMetadataTag(exhaustTag);
    }

    try {
        const safeLocalId = builder.build();
        console.log(`   Safe building result: ${safeLocalId}`);
    } catch (e: any) {
        console.log(`   Safe building failed: ${e.message}`);
    }

    // 6. Verbose mode demonstration
    console.log("\n6. Verbose Mode...");

    const verboseLocalId = LocalIdBuilder.create(version)
        .withVerboseMode(true)
        .withPrimaryItem(primaryPath)
        .withMetadataTag(
            codebooks.createTag(CodebookName.Quantity, "temperature"),
        )
        .build();

    const regularLocalId = LocalIdBuilder.create(version)
        .withVerboseMode(false)
        .withPrimaryItem(primaryPath)
        .withMetadataTag(
            codebooks.createTag(CodebookName.Quantity, "temperature"),
        )
        .build();

    console.log(`   Verbose mode: ${verboseLocalId}`);
    console.log(`   Regular mode: ${regularLocalId}`);

    console.log("\n=== Advanced operations completed! ===");
}

main().catch(console.error);
